### DISCUSSION TITLE
Evaluating dispersion curves for phononic crystals using MOOSE

### DISCUSSION THREAD

**Discussion by adhithya96** [idaholab/moose#27669] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello, 

I have worked on a small code in C++ to evaluate the dispersion diagram for a square plate. This includes implementing the Bloch boundary condition and solving the eigenvalue problem. I have used Eigen library to do this. This following is the github link to my code. [Bloch2D](https://github.com/adhithya96/EigenSolverMOOSE/tree/main/Bloch2D) 
For the past few months, I have been trying to implement the same in MOOSE but was unable to navigate through MOOSE codebase. I don't think I can implement this without help from MOOSE team. Is this something that you are looking for ?

**Comment by GiudGiud** [comment]
@laagesen @dschwen

**Reply by laagesen** [reply]
Hello, this is something that might be useful for a pending project that we put in an application for funding on. Do you have a paper, book or other reference that describes the governing equations?

**Reply by adhithya96** [reply]
Yes. This is the paper I used to implement the code.
[Implementation of Bloch boundary condition.pdf](https://github.com/idaholab/moose/files/15379227/Implementation.of.Bloch.boundary.condition.pdf)
For validation I used the test case from this paper.
[Validation.pdf](https://github.com/idaholab/moose/files/15379259/Validation.pdf)

**Reply by laagesen** [reply]
Have you checked out the eigenvalue problem documentation in MOOSE already? 

https://mooseframework.inl.gov/source/problems/EigenProblem.html

This is outside my area of expertise but seems like a good starting point. Have a look and if this doesn't answer your question we may want to contact one of the people who works on reactor physics, which is what this capability was developed for.

**Reply by adhithya96** [reply]
Yes. I spent around two months trying to figure out how to implement it in MOOSE. 
Couple of places where I was not able to figure out how to solve the problem are 
1. The eigenvalue problem is used for diffusion equation while the problem is a vector valued function with plane strain assumption for 2D. I wasn't sure if the tensor mechanics module was coupled with the eigenvalue problem. 
2. Implementation using this methodology required me to modify the global mass and stiffness matrix by multiplying it with constraint matrix. I couldn't find a way to access the global matrices. Even in the MOOSE code the element matrices were added to the appropriate entries in the global matrices. 
3. The mesh is duplicated for taking the real and imaginary parts. I tried to mimic this by considering four dofs in one node but the assembly for this case is different from the normal fe assembly because of the imaginary components. 
4. The global mass and stiffness matrix needs to be constructed only once. Only the constraint matrix depends on the wave number which is varied to get the reduced matrices and get eigenvalues.

**Comment by laagesen** [comment]
For #1, I think the reactor physics people use this for groups of neutrons with different energies, so there is probably a solution. Would someone from that area be able to help here, maybe @loganharbour or @YaqiWang or @jortensi ?

**Reply by adhithya96** [reply]
I am extending the above code to evaluate dispersion curves for phononic crystals with holes as well using XFEM. 
It would be nice to contribute to the MOOSE framework.
